zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1945,19 +1941,17 @@ public void asyncCreate(final String path, Object datat, final CreateMode mode,
   }
 
   private void doAsyncCreate(final String path, final byte[] data, final CreateMode mode,
-      final long startT, final ZkAsyncCallbacks.CreateCallbackHandler cb,
-      final String expectedSessionId) {
+      final long startT, final ZkAsyncCallbacks.CreateCallbackHandler cb, final String expectedSessionId) {
     try {
       retryUntilConnected(() -> {
-        getExpectedZookeeper(expectedSessionId)
-            .create(path, data, ZooDefs.Ids.OPEN_ACL_UNSAFE, mode, cb,
-                new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false) {
-                  @Override
-                  protected void doRetry() {
-                    doAsyncCreate(path, data, mode, System.currentTimeMillis(), cb,
-                        expectedSessionId);
-                  }
-                });
+        getExpectedZookeeper(expectedSessionId).create(path, data, ZooDefs.Ids.OPEN_ACL_UNSAFE, mode, cb,
+            new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, false,
+                GZipCompressionUtil.isCompressed(data)) {
```

Comment:
A question on the design:

I could see an alternative code path where the metric is incremented inside `GZipCompressionUtil` (it would be nice if we had an interface (say `ICompressionUtil` for example) and `GZipCompressionUtil` was an implementation) whenever compression is actually done? Or is it just as hard as doing it to the serializer interface?

Also what kind of value does this really add? Curious because I thought we were already emitting metrics for oversized ZNodes. 

Also, I'm not sure if I'm comfortable creating a metric that could potentially be inaccurate - could you explain why this might not be accurate? Are there zkClient write paths we're not covering?

Reply:
If the metric is in GZipCompressionUtil, it would be impossible for us to identify the caller. Given the util can be used in more than one ZkClients, which ZkClientMonitor should be added in? Or, if we are talking about a new monitor which reports JVM level node compression statistics, then this metric may not be able to provide the necessary insight when we have multiple Helix instances in one JVM. For example, if there are multiple controllers managing different clusters in one JVM, we don't know which cluster contains the compressed ZK node.
Another problem of monitoring in the util is that we never know if the write operation is eventually done or not. For instance, after compression, the node is still oversized. And ZkClient fails that operation. But we will still plus one to the metric.
All the arguments above convince me that adding the report logic to the ZkClient, immediately after write request returns, is the most accurate and possible way.

> emitting metrics for oversized ZNodes
Which metrics are you talking about? I assume you are talking about the ZK metric that counts large nodes based on scan? This may not be sufficient. 1. It reports based on a snapshot when the scan happens. If a node is temporarily oversized, then it may not be caught by the scan. 2. Fundamentally, the oversized node is a different concern. Most of the oversized nodes are much smaller after compression. And oversized node scan won't report them. In addition, with the recent change that we allow users to set a lower threshold for compression, compression can happen even the node is small. 

Regarding inaccuracy, if the user uses a byte serializer and compresses data (in the application code) before passing it to a Helix ZkClient, the metric will treat the request the same as a Helix compressed node and increase the count. This is a corner case, but worth pointing out. But I don't think it would be a major concern.